src/include/daos/checksum.h

```diff
@@ -15,48 +15,259 @@
+ *				for the extents \a recxs. The total data
+ *				length of the sgl should be the same as the sum
+ *				of the lengths of all recxs
+ * @param[in]	rec_len		number of bytes for each record
+ * @param[in]	recxs		record extents
+ * @param[in]	nr		number of record extents and number of
+ *				csum_arrays that will be allocated
+ * @param[out]	pcsum_bufs	csum_bufs for the checksums created for each
+ *				extent
+ *
+ * @return			0 for success, or an error code
+ */
+int
+daos_csummer_calc_csum(struct daos_csummer *obj, d_sg_list_t *sgl,
+		       size_t rec_len, daos_recx_t *recxs, size_t nr,
+		       daos_csum_buf_t **pcsum_bufs);
+
+/**
+ * Allocate memory for the daos_csum_buf_t structures and the memory buffer for
+ * each csum within the structure. Will initialize the daos_csum_buf_t to
+ * appropriate values.
+ * @param[in]	obj		the daos_csummer object
+ * @param[in]	rec_len		record length
+ * @param[in]	nr		number of record extents
+ * @param[in]	recxs		record extents
+ * @param[out]	pcsum_bufs	csum_bufs for the checksums created for each
+ *				extent
+ *
+ * @return			0 for success, or an error code
+ */
+int
+daos_csummer_prep_csum_buf(struct daos_csummer *obj, size_t rec_len, size_t nr,
+			   daos_recx_t *recxs, daos_csum_buf_t **pcsum_bufs);
+
+/** Destroy the csum buf and memory allocated for checksums */
+void
+daos_csummer_destroy_csum_buf(struct daos_csummer *obj,
+			      daos_csum_buf_t **pcsum_buf);
+
+/**
+ * -----------------------------------------------------------------------------
+ * daos_csum_buf_t Functions
+ * -----------------------------------------------------------------------------
+ */
+
+/** Setup the \a daos_csum_buf_t with the checksum buffer,
+ * buffer size, csum size, number of checksums, and the chunksize the
+ * checksum represents
+ */
+void
+dcb_set(daos_csum_buf_t *csum_buf, void *buf,
+	uint32_t csum_buf_size,
+	uint16_t csum_size,
+	uint32_t csum_count,
+	uint32_t chunksize);
+
+/** Set the csum buf to a NULL value */
+void
+dcb_set_null(daos_csum_buf_t *csum_buf);
+
+/** Is the \a daos_csum_buf_t setup appropriately to be used */
+bool
+dcb_is_valid(const daos_csum_buf_t *csum);
+
+
+/** Returns the index of the checksum provided the offset into the data
+ * the checksums are derived from.
+ */
+uint32_t
+dcb_off2idx(daos_csum_buf_t *csum_buf, uint32_t offset_bytes);
+
+/** Returns the appropriate checksum given the index of the checksum. */
+uint8_t *
+dcb_idx2csum(daos_csum_buf_t *csum_buf, uint32_t idx);
+
+/** Returns the appropriate checksum given the data offset based on chunk
+ *	information.
+ */
+uint8_t *
+dcb_off2csum(daos_csum_buf_t *csum_buf, uint32_t offset);
+
+
+/** Gets the number checksums needed given a record extent.  */
+uint32_t
+daos_recx_calc_chunks(daos_recx_t extent, uint32_t record_size,
+		      uint32_t chunk_size);
+
+
+/** Helper function for dividing a range (lo-hi) into number of chunks, using
+ * absolute alignment
+ */
```

Comment:
Does this function need to be externally visible? It might be separate out externally visible and internally visible functions in separate files?

Reply:
External to DAOS? No I don't think it should be needed externally. It is different than the other functions, however, in that it doesn't operate on a DAOS specific type so maybe it belongs in a utilities/helpers file? I was hesitant to create a new file for a single function, however. Do you have a suggestion?